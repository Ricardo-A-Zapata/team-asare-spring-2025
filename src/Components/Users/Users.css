/* User Form Styles */
.form-container {
  background-color: white;
  border-radius: 8px;
  box-shadow: 0 4px 16px rgba(0, 0, 0, 0.15);
  padding: 2rem;
  margin-bottom: 2rem;
  max-width: 500px;
  margin-left: auto;
  margin-right: auto;
  position: relative;
  animation: fadeIn 0.3s ease-in-out;
}

@keyframes fadeIn {
  from {
    opacity: 0;
    transform: translateY(-10px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.user-form {
  width: 100%;
}

.form-container h3 {
  margin-top: 0;
  margin-bottom: 1.5rem;
  color: #333;
  font-size: 1.5rem;
  border-bottom: 1px solid #eee;
  padding-bottom: 0.75rem;
}

/* Override the default form label styles completely */
.form-container form label {
  position: static;
  height: auto;
  width: auto;
  margin-left: 0;
  font-weight: 600;
  font-size: 0.9rem;
  color: #444;
  margin-bottom: 6px;
  display: block;
  background: none;
  border: none;
}

/* Remove the pseudo elements that create the label boxes */
.form-container form label::before,
.form-container form label::after {
  display: none;
}

.form-fields {
  margin-bottom: 1.5rem;
}

.form-field {
  margin-bottom: 1.2rem;
}

/* Override the default form input styles completely */
.form-container input[type="text"],
.form-container input[type="email"],
.form-container input[disabled] {
  display: block;
  width: 100%;
  height: 38px;
  padding: 6px 12px;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: 0.9rem;
  background-color: #fff;
  margin: 0;
  max-width: 100%;
  box-sizing: border-box;
}

.form-container input[disabled] {
  background-color: #f5f5f5;
  color: #777;
  cursor: not-allowed;
}

.form-container input:focus {
  border-color: #0066cc;
  box-shadow: 0 0 0 2px rgba(0, 102, 204, 0.15);
  outline: none;
}

/* Roles section */
.roles-container {
  margin: 20px 0;
  padding: 15px;
  background-color: #f8f9fa;
  border: 1px solid #eaeaea;
  border-radius: 5px;
}

.roles-container h4 {
  margin-top: 0;
  margin-bottom: 12px;
  font-size: 1rem;
  color: #444;
  font-weight: 600;
}

.roles-checkboxes {
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
}

/* Override default styles for checkboxes */
.role-checkbox {
  display: flex;
  align-items: center;
  padding: 8px 12px;
  background-color: white;
  border: 1px solid #ddd;
  border-radius: 4px;
  cursor: pointer;
  transition: background-color 0.2s, border-color 0.2s;
}

.role-checkbox:hover {
  background-color: #f0f7ff;
  border-color: #0066cc;
}

.role-checkbox input[type="checkbox"] {
  margin: 0;
  margin-right: 8px;
  height: 16px;
  width: 16px;
  cursor: pointer;
}

.role-checkbox label {
  margin: 0 !important;
  padding: 0 !important;
  height: auto !important;
  position: static !important;
  font-weight: normal !important;
  font-size: 0.9rem !important;
  cursor: pointer;
}

/* Form actions */
.form-actions {
  display: flex;
  justify-content: flex-end;
  gap: 10px;
  margin-top: 25px;
}

.form-actions button {
  padding: 8px 16px;
  font-size: 0.9rem;
  border-radius: 4px;
  cursor: pointer;
  height: auto;
  width: auto;
  transition: all 0.2s ease;
}

.form-actions button[type="button"] {
  background-color: #f2f2f2;
  border: 1px solid #ccc;
  color: #333;
}

.form-actions button[type="button"]:hover {
  background-color: #e5e5e5;
}

.form-actions button[type="submit"] {
  background-color: #0066cc;
  border: 1px solid #0055aa;
  color: white;
  font-weight: 500;
}

.form-actions button[type="submit"]:hover {
  background-color: #0055aa;
} 