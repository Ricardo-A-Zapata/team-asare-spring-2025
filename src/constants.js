export const BACKEND_URL = process.env.REACT_APP_URL_PRE || 'https://teamasare.pythonanywhere.com';

// Flag to determine if we should use direct API calls or a development proxy
export const USE_DEV_PROXY = process.env.REACT_APP_USE_PROXY === 'true' || false;

// Helper function to construct API URLs with potential proxy prefix
export const getApiUrl = (endpoint) => {
  // Remove trailing slash from backend URL if present
  const baseUrl = BACKEND_URL.endsWith('/') ? BACKEND_URL.slice(0, -1) : BACKEND_URL;
  
  // Add leading slash to endpoint if not present
  const formattedEndpoint = endpoint.startsWith('/') ? endpoint : `/${endpoint}`;
  
  // When in development with proxy enabled, use relative URLs
  if (USE_DEV_PROXY && process.env.NODE_ENV === 'development') {
    return formattedEndpoint;
  }
  
  // Otherwise use the full URL
  return `${baseUrl}${formattedEndpoint}`;
};

// Manuscript states
export const MANUSCRIPT_STATES = {
  SUBMITTED: 'SUBMITTED',
  REJECTED: 'REJECTED',
  REFEREE_REVIEW: 'REFEREE_REVIEW',
  AUTHOR_REVISIONS: 'AUTHOR_REVISIONS',
  WITHDRAWN: 'WITHDRAWN',
  COPY_EDIT: 'COPY_EDIT',
  AUTHOR_REVIEW: 'AUTHOR_REVIEW',
  FORMATTING: 'FORMATTING',
  PUBLISHED: 'PUBLISHED',
  EDITOR_REVIEW: 'EDITOR_REVIEW'
};

// Verdict types
export const VERDICT_TYPES = {
  ACCEPT: 'ACCEPT',
  REJECT: 'REJECT',
  ACCEPT_WITH_REVISIONS: 'ACCEPT_WITH_REVISIONS',
  MINOR_REVISIONS: 'MINOR_REVISIONS',
  MAJOR_REVISIONS: 'MAJOR_REVISIONS'
};

// User roles
export const USER_ROLES = {
  EDITOR: 'ED',
  REFEREE: 'RE',
  AUTHOR: 'AU'
};

// API endpoints
export const API_ENDPOINTS = {
  MANUSCRIPTS_READ: '/manuscripts',
  USERS_READ: '/user/read',
  MANUSCRIPT_STATE: '/manuscript/state',
  MANUSCRIPT_REFEREE: '/manuscript/referee',
  MANUSCRIPT_REVIEW: '/manuscript/review',
  MANUSCRIPT_WITHDRAW: '/manuscript/withdraw',
  MANUSCRIPT_TEXT: '/manuscript/text',
  MANUSCRIPT_CREATE: '/manuscript/create'
};

// Local storage keys
export const STORAGE_KEYS = {
  LOGGED_IN: 'loggedIn',
  EMAIL: 'email',
  MANUSCRIPT_REVIEW: 'manuscript_review_'
};
