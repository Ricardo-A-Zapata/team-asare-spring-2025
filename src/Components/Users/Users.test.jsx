import React from 'react';
import { render, screen, waitFor, act, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Users, { filterUsers, sortUsers } from './Users';
import axios from 'axios';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

jest.mock('axios');

import { BACKEND_URL } from '../../constants';

const USERS_READ_ENDPOINT = `${BACKEND_URL}/user/read`;
const USER_CREATE_ENDPOINT = `${BACKEND_URL}/user/create`;
const USER_UPDATE_ENDPOINT = `${BACKEND_URL}/user/update`;
const USER_DELETE_ENDPOINT = `${BACKEND_URL}/user/delete`;

// Helper function to wrap component with Router
const renderWithRouter = (ui) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>);
};

// Test data
const testUsers = {
  'user1@example.com': {
    name: 'John Doe',
    email: 'user1@example.com',
    affiliation: 'University A',
    roleCodes: ['AU', 'RE']
  },
  'user2@example.com': {
    name: 'Jane Smith',
    email: 'user2@example.com',
    affiliation: 'Company B',
    roleCodes: ['ED']
  },
  'user3@example.com': {
    name: 'Bob Johnson',
    email: 'user3@example.com',
    affiliation: 'Organization C',
    roleCodes: ['AU']
  }
};

describe('sortUsers function', () => {
  const testUserArray = Object.values(testUsers);
  
  it('returns the original array if no sortConfig is provided', () => {
    const result = sortUsers(testUserArray, null);
    expect(result).toEqual(testUserArray);
  });
  
  it('sorts by name in ascending order', () => {
    const result = sortUsers(testUserArray, { key: 'name', direction: 'asc' });
    expect(result[0].name).toBe('Bob Johnson');
    expect(result[1].name).toBe('Jane Smith');
    expect(result[2].name).toBe('John Doe');
  });
  
  it('sorts by name in descending order', () => {
    const result = sortUsers(testUserArray, { key: 'name', direction: 'desc' });
    expect(result[0].name).toBe('John Doe');
    expect(result[1].name).toBe('Jane Smith');
    expect(result[2].name).toBe('Bob Johnson');
  });
  
  it('sorts by email in ascending order', () => {
    const result = sortUsers(testUserArray, { key: 'email', direction: 'asc' });
    expect(result[0].email).toBe('user1@example.com');
    expect(result[1].email).toBe('user2@example.com');
    expect(result[2].email).toBe('user3@example.com');
  });
  
  it('sorts by affiliation in ascending order', () => {
    const result = sortUsers(testUserArray, { key: 'affiliation', direction: 'asc' });
    expect(result[0].affiliation).toBe('Company B');
    expect(result[1].affiliation).toBe('Organization C');
    expect(result[2].affiliation).toBe('University A');
  });
  
  it('sorts by role in ascending order', () => {
    const result = sortUsers(testUserArray, { key: 'role', direction: 'asc' });
    // First role of Bob and John is 'Author', Jane's is 'Editor'
    // Alphabetically: Author, Author, Editor
    const firstRoles = result.map(user => 
      user.roleCodes && user.roleCodes.length > 0 ? user.roleCodes[0] : ''
    );
    expect(firstRoles[0]).toBe('AU');
    expect(firstRoles[1]).toBe('AU');
    expect(firstRoles[2]).toBe('ED');
  });
  
  it('handles missing affiliation values', () => {
    const usersWithMissingData = [
      ...testUserArray,
      { name: 'Alice', email: 'alice@example.com', roleCodes: ['RE'] } // No affiliation
    ];
    
    const result = sortUsers(usersWithMissingData, { key: 'affiliation', direction: 'asc' });
    expect(result[0].name).toBe('Alice'); // Empty affiliation comes first alphabetically
  });
});

describe('Users Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    axios.get.mockResolvedValue({
      data: {
        Users: testUsers
      }
    });
    axios.put.mockResolvedValue({});
    axios.delete.mockResolvedValue({});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the component and loads users', async () => {
    renderWithRouter(<Users />);
    
    // Check for heading
    expect(screen.getByText('View All Users')).toBeInTheDocument();
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
      expect(screen.getByText('Jane Smith')).toBeInTheDocument();
      expect(screen.getByText('Bob Johnson')).toBeInTheDocument();
    });
  });

  it('filters users by search term', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Get the search input and type a search term
    const searchInput = screen.getByPlaceholderText('Search by name or email');
    fireEvent.change(searchInput, { target: { value: 'Jane' } });
    
    // Check that only Jane is displayed
    expect(screen.queryByText('John Doe')).not.toBeInTheDocument();
    expect(screen.getByText('Jane Smith')).toBeInTheDocument();
    expect(screen.queryByText('Bob Johnson')).not.toBeInTheDocument();
    
    // Clear search
    fireEvent.change(searchInput, { target: { value: '' } });
    
    // Check that all users are displayed again
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
      expect(screen.getByText('Jane Smith')).toBeInTheDocument();
      expect(screen.getByText('Bob Johnson')).toBeInTheDocument();
    });
  });

  it('filters users by role', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Select Editor role
    const roleSelect = screen.getByLabelText('Filter by role');
    fireEvent.change(roleSelect, { target: { value: 'ED' } });
    
    // Check that only Jane is displayed (she's the only Editor)
    expect(screen.queryByText('John Doe')).not.toBeInTheDocument();
    expect(screen.getByText('Jane Smith')).toBeInTheDocument();
    expect(screen.queryByText('Bob Johnson')).not.toBeInTheDocument();
    
    // Select Author role
    fireEvent.change(roleSelect, { target: { value: 'AU' } });
    
    // Check that only John and Bob are displayed (they're the Authors)
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.queryByText('Jane Smith')).not.toBeInTheDocument();
    expect(screen.getByText('Bob Johnson')).toBeInTheDocument();
  });

  it('shows "no results" message when filters match no users', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Search for a non-existent user
    const searchInput = screen.getByPlaceholderText('Search by name or email');
    fireEvent.change(searchInput, { target: { value: 'NonExistentUser' } });
    
    // Check that the no results message is displayed
    expect(screen.getByText('No users match your filters. Try adjusting your search criteria.')).toBeInTheDocument();
  });

  it('clears filters when clear button is clicked', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Apply a filter
    const searchInput = screen.getByPlaceholderText('Search by name or email');
    fireEvent.change(searchInput, { target: { value: 'Jane' } });
    
    // Check that only Jane is displayed
    expect(screen.queryByText('John Doe')).not.toBeInTheDocument();
    expect(screen.getByText('Jane Smith')).toBeInTheDocument();
    
    // Click the clear filters button
    const clearButton = screen.getByText('Clear Filters');
    fireEvent.click(clearButton);
    
    // Check that all users are displayed again
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
      expect(screen.getByText('Jane Smith')).toBeInTheDocument();
      expect(screen.getByText('Bob Johnson')).toBeInTheDocument();
    });
  });

  test('renders user links with correct href attributes', async () => {
    const mockUsers = {
      Users: [{ id: 'john@example.com', name: 'John Doe', email: 'john@example.com' }]
    };

    axios.get.mockResolvedValueOnce({ data: mockUsers });

    await act(async () => {
      render(
        <BrowserRouter>
          <Users />
        </BrowserRouter>
      );
    });

    await waitFor(() => {
      const userLink = screen.getByRole('link', { name: /John Doe/i });
      expect(userLink).toHaveAttribute('href', `/john@example.com`);
    });
  });

  test('matches snapshot', async () => {
    let asFragment;
    
    await act(async () => {
      const rendered = render(
        <BrowserRouter>
          <Users />
        </BrowserRouter>
      );
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test('opens AddUserForm when clicking "Add a User" button', async () => {
    axios.get.mockResolvedValueOnce({ data: { Users: [] } });
    
    await act(async () => {
      render(
        <BrowserRouter>
          <Users />
        </BrowserRouter>
      );
    });

    expect(screen.queryByText(/Add New User/i)).not.toBeInTheDocument();
    const addUserButton = screen.getByRole('button', { name: /Add a User/i });
    await userEvent.click(addUserButton);
    expect(screen.getByText(/Add New User/i)).toBeInTheDocument();
  });

  test('opens EditUserForm when clicking "Edit" button', async () => {
    const mockUsers = {
      Users: [
        { id: 'test@gmail.com', name: 'test', email: 'test@gmail.com', affiliation: 'nyu' }
      ]
    };
    axios.get.mockResolvedValueOnce({ data: mockUsers });
    
    await act(async () => {
      render(
        <BrowserRouter>
          <Users />
        </BrowserRouter>
      );
    });
    const editButton = screen.getByRole('button', { name: /Edit/i });
    await userEvent.click(editButton);
    expect(screen.getByText((content, element) =>
        content.includes("Update") && content.includes("User"))
    ).toBeInTheDocument();
  });

  it('sorts users when selecting a sort option', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Get the sort dropdown and select "Name (Z-A)"
    const sortSelect = screen.getByLabelText('Sort users');
    fireEvent.change(sortSelect, { target: { value: 'name-desc' } });
    
    // Check that users are sorted by name in descending order
    const userElements = screen.getAllByRole('heading', { level: 2 });
    expect(userElements[0].textContent).toBe('John Doe');
    
    // Change sort to "Name (A-Z)"
    fireEvent.change(sortSelect, { target: { value: 'name-asc' } });
    
    // Check that users are now sorted in ascending order
    const userElementsAfterSort = screen.getAllByRole('heading', { level: 2 });
    expect(userElementsAfterSort[0].textContent).toBe('Bob Johnson');
  });

  it('maintains sort order when applying filters', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Set sort to "Name (A-Z)"
    const sortSelect = screen.getByLabelText('Sort users');
    fireEvent.change(sortSelect, { target: { value: 'name-asc' } });
    
    // Apply a role filter for "Author"
    const roleSelect = screen.getByLabelText('Filter by role');
    fireEvent.change(roleSelect, { target: { value: 'AU' } });
    
    // Should show Bob Johnson first (alphabetical) and John Doe second
    // Both are authors
    const userElements = screen.getAllByRole('heading', { level: 2 });
    expect(userElements[0].textContent).toBe('Bob Johnson');
    expect(userElements[1].textContent).toBe('John Doe');
    expect(screen.queryByText('Jane Smith')).not.toBeInTheDocument();
  });

  it('clears sorting when selecting default option', async () => {
    renderWithRouter(<Users />);
    
    // Wait for users to load
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
    
    // Set sort to "Name (A-Z)"
    const sortSelect = screen.getByLabelText('Sort users');
    fireEvent.change(sortSelect, { target: { value: 'name-asc' } });
    
    // Verify sort applied
    const userElementsSorted = screen.getAllByRole('heading', { level: 2 });
    expect(userElementsSorted[0].textContent).toBe('Bob Johnson');
    
    // Clear sorting
    fireEvent.change(sortSelect, { target: { value: '' } });
    
    // Verify original order is restored (or at least it's not sorted by name asc anymore)
    // This test is a bit tricky as the original order depends on how the users are stored in the state
    const userElementsUnsorted = screen.getAllByRole('heading', { level: 2 });
    expect(userElementsUnsorted.length).toBe(3); // All users should be visible
  });
});
