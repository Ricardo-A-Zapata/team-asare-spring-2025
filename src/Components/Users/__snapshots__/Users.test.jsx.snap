// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <header>
      <h1>
        View All Users
      </h1>
      <button
        type="button"
      >
        Add a User
      </button>
    </header>
    <div
      class="user-controls"
    >
      <div
        class="user-filters"
      >
        <div
          class="search-bar"
        >
          <input
            placeholder="Search by name or email"
            type="text"
            value=""
          />
        </div>
        <div
          class="role-filter"
        >
          <select
            aria-label="Filter by role"
          >
            <option
              value=""
            >
              All Roles
            </option>
          </select>
        </div>
      </div>
      <div
        class="user-sorting"
      >
        <select
          aria-label="Sort users"
        >
          <option
            value=""
          >
            Default Order
          </option>
          <option
            value="name-asc"
          >
            Name (A-Z)
          </option>
          <option
            value="name-desc"
          >
            Name (Z-A)
          </option>
          <option
            value="email-asc"
          >
            Email (A-Z)
          </option>
          <option
            value="email-desc"
          >
            Email (Z-A)
          </option>
          <option
            value="affiliation-asc"
          >
            Affiliation (A-Z)
          </option>
          <option
            value="affiliation-desc"
          >
            Affiliation (Z-A)
          </option>
          <option
            value="role-asc"
          >
            Role (A-Z)
          </option>
          <option
            value="role-desc"
          >
            Role (Z-A)
          </option>
        </select>
      </div>
    </div>
    <div
      class="user-count"
    >
      Showing all 3 users
    </div>
    <div
      class="error-message"
    >
      There was a problem retrieving the roles. Error: Unknown endpoint: https://teamasare.pythonanywhere.com//roles
    </div>
    <div
      class="user-container"
    >
      <a
        href="/user1@example.com"
      >
        <h2>
          John Doe
        </h2>
        <p>
          Email: user1@example.com
        </p>
        <p>
          Affiliation: University A
        </p>
        <p>
          Roles: AU, RE
        </p>
      </a>
      <div
        class="button-group"
      >
        <button
          type="button"
        >
          Delete
        </button>
        <button
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
    <div
      class="user-container"
    >
      <a
        href="/user2@example.com"
      >
        <h2>
          Jane Smith
        </h2>
        <p>
          Email: user2@example.com
        </p>
        <p>
          Affiliation: Company B
        </p>
        <p>
          Roles: ED
        </p>
      </a>
      <div
        class="button-group"
      >
        <button
          type="button"
        >
          Delete
        </button>
        <button
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
    <div
      class="user-container"
    >
      <a
        href="/user3@example.com"
      >
        <h2>
          Bob Johnson
        </h2>
        <p>
          Email: user3@example.com
        </p>
        <p>
          Affiliation: Organization C
        </p>
        <p>
          Roles: AU
        </p>
      </a>
      <div
        class="button-group"
      >
        <button
          type="button"
        >
          Delete
        </button>
        <button
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
